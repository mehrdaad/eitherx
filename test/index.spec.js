import React from 'react'
import renderer from 'react-test-renderer'

// Hack to use `src` when running coverage
const Eitherx = process.env.npm_package_scripts_test_coverage
  ? require('../src').default
  : require('../dist/index.js')

test('should throw if < 1 or > 2 children are passed', () => {
  expect(() => renderer.create(<Eitherx />)).toThrowErrorMatchingSnapshot()
  expect(() =>
    renderer.create(
      <Eitherx>
        <div />
        <div />
        <div />
      </Eitherx>
    )
  ).toThrowErrorMatchingSnapshot()
})

test('should throw render is passed but not catchError', () => {
  expect(() => renderer.create(<Eitherx render={() => <div />} />)).toThrowErrorMatchingSnapshot()
})

test('should throw if render or catchError is not a function', () => {
  expect(() =>
    renderer.create(<Eitherx render={() => <div />} catchError={'Not a function'} />)
  ).toThrowErrorMatchingSnapshot()

  expect(() =>
    renderer.create(<Eitherx render={'Not a function'} catchError={() => <div />} />)
  ).toThrowErrorMatchingSnapshot()
})

class ErrorComponent extends React.Component {
  constructor(props) {
    super(props)
  }
  render() {
    throw new Error('Test Error')
  }
}

test('should return null if no error handler is passed, and error is thrown', () => {
  let component = renderer.create(
    <Eitherx>
      <p>My Paragraph</p>
    </Eitherx>
  )
  expect(component.toJSON()).toMatchSnapshot()

  component = renderer.create(
    <Eitherx>
      <ErrorComponent>"My Error"</ErrorComponent>
    </Eitherx>
  )
  expect(component.toJSON()).toEqual(null)
})

test('should render second component on error', () => {
  let component = renderer.create(
    <Eitherx>
      <p>"Happy Child üòÑ"</p>
      <p>"Error as child ‚òπÔ∏è"</p>
    </Eitherx>
  )
  expect(component.toJSON()).toMatchSnapshot()

  component = renderer.create(
    <Eitherx>
      <ErrorComponent>"Function as child üòÑ"</ErrorComponent>
      <p>"Error as child ‚òπÔ∏è"</p>
    </Eitherx>
  )
  expect(component.toJSON()).toMatchSnapshot()
})

test('should work with render-catch props as well', () => {
  let component = renderer.create(
    <Eitherx render={() => <p>"Render Prop"</p>} catchError={() => <p>"Catch Prop"</p>} />
  )
  expect(component.toJSON()).toMatchSnapshot()

  component = renderer.create(
    <Eitherx
      render={() => <ErrorComponent>"Render Prop"</ErrorComponent>}
      catchError={() => <p>"Catch Prop"</p>}
    />
  )
  expect(component.toJSON()).toMatchSnapshot()
})

test('should allow further error handling with `handleError`', () => {
  const handleError = jest.fn(() => true)
  const catchError = jest.fn(() => <p>"Catch Prop"</p>)

  let component = renderer.create(
    <Eitherx
      handleError={handleError}
      render={() => <ErrorComponent>"Render Prop"</ErrorComponent>}
      catchError={catchError}
    />
  )

  expect(handleError).toHaveBeenCalledWith(
    expect.objectContaining({
      error: expect.any(Object),
      info: expect.any(Object)
    })
  )
  expect(catchError).toHaveBeenCalledWith(
    expect.objectContaining({
      error: expect.any(Object)
    })
  )

  expect(component).toMatchSnapshot()
})
