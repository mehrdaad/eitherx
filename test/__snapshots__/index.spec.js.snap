// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow further error handling with \`handleError\` 1`] = `
<p>
  "Catch Prop"
</p>
`;

exports[`should render second component on error 1`] = `
<p>
  "Happy Child üòÑ"
</p>
`;

exports[`should render second component on error 2`] = `
<p>
  "Error as child ‚òπÔ∏è"
</p>
`;

exports[`should return null if no error handler is passed, and error is thrown 1`] = `
<p>
  My Paragraph
</p>
`;

exports[`should throw if < 1 or > 2 children are passed 1`] = `"Eitherx expects either 1 or 2 children. You passed 0 children"`;

exports[`should throw if < 1 or > 2 children are passed 2`] = `"Eitherx expects either 1 or 2 children. You passed 3 children"`;

exports[`should throw if render or catchError is not a function 1`] = `"\`render\` and \`catchError\` must both be functions"`;

exports[`should throw if render or catchError is not a function 2`] = `"\`render\` and \`catchError\` must both be functions"`;

exports[`should throw render is passed but not catchError 1`] = `"If \`render\` prop is passed, you must also pass a \`catchError\` callback"`;

exports[`should work with render-catch props as well 1`] = `
<p>
  "Render Prop"
</p>
`;

exports[`should work with render-catch props as well 2`] = `
<p>
  "Catch Prop"
</p>
`;
